File: backend/app/engine/units.py
"""Units, rounding (IS 1200) and the core result types of the engine.

Every quantity the engine produces is a :class:`Quantity` carrying an audit
trail of :class:`FormulaStep` objects, so any number can be explained back to
its formula, inputs and IS clause. See project.md section 6.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

MM_PER_M = 1000.0


def mm_to_m(value_mm: float | None) -> float:
    """Convert a millimetre dimension (as read from a drawing) to metres."""
    return (value_mm or 0.0) / MM_PER_M


# Presentation rounding per IS 1200 conventions. Raw values are kept full
# precision internally; rounding is applied only when presenting a line item,
# and the rule is recorded so totals stay reproducible.
ROUNDING = {
    "m": 2,      # linear
    "m2": 2,     # area
    "m3": 3,     # volume (DSR often 2; engine keeps 3, UI may show 2)
    "kg": 2,     # steel weight
    "MT": 3,
    "Nos": 0,
}


def round_qty(value: float, unit: str) -> float:
    return round(value, ROUNDING.get(unit, 3))


@dataclass
class FormulaStep:
    """One auditable step in a calculation."""

    formula_id: str            # e.g. "concrete.column.volume"
    expression: str            # human-readable, e.g. "b*D*H*count"
    inputs: dict[str, Any]     # {"b_m": 0.3, "D_m": 0.6, ...}
    result: float
    clause_ref: str = ""       # e.g. "IS 1200 Part 2"

    def to_dict(self) -> dict[str, Any]:
        return {
            "formula_id": self.formula_id,
            "expression": self.expression,
            "inputs": self.inputs,
            "result": self.result,
            "clause_ref": self.clause_ref,
        }


@dataclass
class Quantity:
    """A single computed BOQ quantity produced from a member."""

    category: str              # concrete | rebar | steel | earthwork | masonry | plaster | formwork
    description: str
    unit: str                  # m3 | m2 | kg | MT | Nos
    value: float               # raw (unrounded) value
    # Traditional measurement-sheet basis columns (No / L / B / D), for the
    # CPWD/DSR-style export. Any may be None.
    nos: float | None = None
    length_m: float | None = None
    breadth_m: float | None = None
    depth_m: float | None = None
    audit: list[FormulaStep] = field(default_factory=list)
    extra: dict[str, Any] = field(default_factory=dict)  # e.g. a BBS table

    def rounded(self) -> float:
        return round_qty(self.value, self.unit)

    def to_dict(self) -> dict[str, Any]:
        return {
            "category": self.category,
            "description": self.description,
            "unit": self.unit,
            "value": self.value,
            "rounded": self.rounded(),
            "nos": self.nos,
            "length_m": self.length_m,
            "breadth_m": self.breadth_m,
            "depth_m": self.depth_m,
            "audit": [s.to_dict() for s in self.audit],
            "extra": self.extra,
        }

File: frontend/src/engine/units.ts
// Units, rounding (IS 1200) and core result types of the in-browser engine.
// Ported from backend/app/engine/units.py — kept numerically identical so the
// static (GitHub Pages) build produces the same quantities as the Python API.

export const MM_PER_M = 1000.0;

export function mmToM(value_mm: number | null | undefined): number {
  return (value_mm || 0.0) / MM_PER_M;
}

// Match Python's round() closely enough for our values: both operate on the
// same IEEE-754 float, so Math.round(x*10^n)/10^n reproduces CPython's result
// for the non-exact-tie products this engine generates.
export function pyRound(x: number, n = 0): number {
  if (!isFinite(x)) return x;
  const m = Math.pow(10, n);
  return Math.round(x * m) / m;
}

// Presentation rounding per IS 1200 conventions.
export const ROUNDING: Record<string, number> = {
  m: 2,
  m2: 2,
  m3: 3,
  kg: 2,
  MT: 3,
  Nos: 0,
};

export function roundQty(value: number, unit: string): number {
  return pyRound(value, unit in ROUNDING ? ROUNDING[unit] : 3);
}

export interface FormulaStep {
  formula_id: string;
  expression: string;
  inputs: Record<string, unknown>;
  result: number;
  clause_ref: string;
}

export function step(
  formula_id: string,
  expression: string,
  inputs: Record<string, unknown>,
  result: number,
  clause_ref = ""
): FormulaStep {
  return { formula_id, expression, inputs, result, clause_ref };
}

export interface Quantity {
  category: string; // concrete | rebar | steel | earthwork | masonry | plaster | formwork
  description: string;
  unit: string; // m3 | m2 | kg | MT | Nos
  value: number; // raw (unrounded)
  nos: number | null;
  length_m: number | null;
  breadth_m: number | null;
  depth_m: number | null;
  audit: FormulaStep[];
  extra: Record<string, any>;
}

export function qty(init: {
  category: string;
  description: string;
  unit: string;
  value: number;
  nos?: number | null;
  length_m?: number | null;
  breadth_m?: number | null;
  depth_m?: number | null;
  audit?: FormulaStep[];
  extra?: Record<string, any>;
}): Quantity {
  return {
    category: init.category,
    description: init.description,
    unit: init.unit,
    value: init.value,
    nos: init.nos ?? null,
    length_m: init.length_m ?? null,
    breadth_m: init.breadth_m ?? null,
    depth_m: init.depth_m ?? null,
    audit: init.audit ?? [],
    extra: init.extra ?? {},
  };
}

// Format a mm dimension like Python's f"{v:.0f}".
export function fmt0(v: number): string {
  return String(Math.round(v));
}
